src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+    where
+        P: IntoIterator,
+        P::Item: ToSql,
+        F: FnOnce(&Row) -> T {
+        if let Some(ref conn) = self.conn {
+            conn.query_row(sql, params, f)
+                .optional()
+                .map_err(|x| Error::new(ErrType::SqliteError(IncomparableError{ err: x })))
+        } else {
+            Err(Error::new(ErrType::SqlConnectionClosed))
+        }
+    }
+}
+
+impl ContractDatabase for SqliteContractDatabase {
+    fn create_map(&mut self, contract_name: &str, map_name: &str, key_type: TupleTypeSignature, value_type: TupleTypeSignature) -> Result<()> {
+        let key_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(key_type));
+        let value_type = TypeSignature::new_atom(AtomTypeIdentifier::TupleType(value_type));
+
+        self.execute("INSERT INTO maps_table (contract_name, map_name, key_type, value_type) VALUES (?, ?, ?, ?)",
+                     &[contract_name, map_name, &key_type.serialize()?, &value_type.serialize()?])?;
+        Ok(())
+    }
+
+    fn fetch_entry(&self, contract_name: &str, map_name: &str, key: &Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), (*key).clone())))
+        }
+
+        let params: [&ToSql; 2] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?];
+
+        let sql_result: Option<Option<String>> = 
+            self.query_row(
+                "SELECT value FROM data_table WHERE map_identifier = ? AND key = ? ORDER BY data_identifier DESC",
+                &params,
+                |row| {
+                    row.get(0)
+                })?;
+        match sql_result {
+            None => {
+                Ok(Value::Void)
+            },
+            Some(sql_result) => {
+                match sql_result {
+                    None => Ok(Value::Void),
+                    Some(value_data) => Value::deserialize(&value_data)
+                }
+            }
+        }
+    }
+
+    fn set_entry(&mut self, contract_name: &str, map_name: &str, key: Value, value: Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(&key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), key)))
+        }
+        if !map_descriptor.value_type.admits(&value) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.value_type), value)))
+        }
+
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &Some(value.serialize()?)];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
```

Comment:
This will fail with a Sqlite error if the values for `(map_identifier, key, value)` already exist.  Did you want `INSERT OR REPLACE`?

Reply:
Actually, it's not a big deal to fix this at this point -- once setting, inserting, and deleting entries is paired with a `Txid`, then everything will effectively be an `INSERT` of a new set of values.